<a href="https://colab.research.google.com/github/klbailey/ADA/blob/main/Copy_of_LAB_19.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#LAB 19

Use train.csv from here https://www.kaggle.com/competitions/titanic/data?select=train.csv. Run a logistic Regression Model following these crucial steps.

1. Replace 'male' with 1 and 'female' with 0 in the 'Sex' column.
2. Have Sex and Age as your Independent Variables (Predictors) and Survived as your DV.

IMPORTANT NOTE: You will encounter this error *"MissingDataError: exog contains inf or nans"* Figure out what it says, figure out how to get past it. The hint is on the error name itself.

Import Libraries

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import statsmodels.api as sm
from sklearn.linear_model import LogisticRegression
import matplotlib.pyplot as plt


Import train.csv into dataframe

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('train.csv')

column_names = df.columns
print(column_names)

# Print data types of each column
data_types = df.dtypes
print("Data types of each column:")
print(data_types)

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')
Data types of each column:
PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object


Check for missing values.

In [ ]:
# Check for missing values
missing_values = df.isnull().sum()
print("Missing values in each column:")
print(missing_values)

Missing values in each column:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


How much is missing? Look at it as percentage.

In [ ]:
# Check for missing values in percentage
missing_percentage = (df.isnull().sum() / len(df)) * 100
print("\nMissing values in percentage:")
print(missing_percentage)


Missing values in percentage:
PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64


Treat 'Embarked' missing values: Embarked column df['Embarked'] Find most common value .mode() Fill in missing value with most common value .fillna(df['Embarked'].mode()[0]) Do this change directly in our existing dataframe inplace=True

In [ ]:
# Fill missing values in 'Embarked' with the mode
df['Embarked'].fillna(df['Embarked'].mode()[0], inplace=True)

Treat 'Cabin' missing values: Due to high percentage of 77.10 we will drop the column.

In [ ]:
# Drop the 'Cabin' column due to a high percentage of missing values
df.drop(columns=['Cabin'], inplace=True)

Treat 'Age' missing values: Fill in with the median.

In [ ]:
# Fill missing values in 'Age' with the median
df['Age'].fillna(df['Age'].median(), inplace=True)


Verify no more missing values.

In [ ]:
# Verify that there are no missing values
print("Missing values after imputation:")
print(df.isnull().sum())

Missing values after imputation:
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


Replace 'male' with 1 and 'female' with 0 in the 'Sex' column.

In [ ]:
# Replace 'male' with 1 and 'female' with 0
df['Sex'] = df['Sex'].replace({'male': 1, 'female': 0})

# Convert the data type of the 'Sex' column to integer
df['Sex'] = df['Sex'].astype(int)

# Check data types of each column again
print("Data types of each column:")
print(df.dtypes)

# Check unique values in the 'Sex' column again
print("Unique values in 'Sex' column:", df['Sex'].unique())


Data types of each column:
PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex              int64
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
dtype: object
Unique values in 'Sex' column: [1 0]


Have Sex and Age as your Independent Variables (Predictors) and Survived as your DV.

In [ ]:
X = df[['Sex', 'Age']]
y = df['Survived']


Fit the logistic regression model.

In [ ]:
import statsmodels.api as sm

# Add a constant term to the independent variables
X = sm.add_constant(X)

# Fit the logistic regression model
model = sm.Logit(y, X).fit()


Optimization terminated successfully.
         Current function value: 0.514730
         Iterations 5


Current function value: 0.514730 is level of fit of the logistic regression model to the data<br>
Iterations 5

Print summary of the model.

In [ ]:
print(model.summary())


                           Logit Regression Results                           
Dep. Variable:               Survived   No. Observations:                  891
Model:                          Logit   Df Residuals:                      888
Method:                           MLE   Df Model:                            2
Date:                Thu, 30 May 2024   Pseudo R-squ.:                  0.2270
Time:                        18:40:21   Log-Likelihood:                -458.62
converged:                       True   LL-Null:                       -593.33
Covariance Type:            nonrobust   LLR p-value:                 3.157e-59
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.1898      0.222      5.361      0.000       0.755       1.625
Sex           -2.5053      0.167    -14.962      0.000      -2.834      -2.177
Age           -0.0047      0.006     -0.743      0.4

DV: Survived

No. of Observations: 891 rows/passengers.

Model: Logit (logistic regression).

Df Residuals: 888 is 891 rows - 3 variables.

Method: MLE Maximum Likelihood Estimation.

Df: 2 degrees of freedom or 2 IV(predictors) in model

Pseudo R-Squared: 0.2270 Sex and Age explain around 22.% of the variation in DV Survived / Sex and Age explain about 22.7% of the variation in whether a passenger survived the Titanic.

Log-Likelihood: -458.62 A less negative log-likelihood suggests the model is better at explaining the observed data.

Converged: The model-building process is smooth and reached a good solution because it is True.

LL-null: -593.33 A lower lL-Null value means the model is better at guessing survival than flipping a coin.

LLR p-Value: 3.157e-59 is a really low score which means sthe model is better than just random guessing.

Coefficients: Show the impact of Sex is -2.5053(male=1) indicating that males are less likely to survive compared to females(0). Age -0.0047(male=1) for every year of age, the likelihood of survival changes by 0.0047 very small so age doesn't seem to have a strong impact on survival.

P>|z|: for both sex and age the p-values are close to zero indicating they are statistically significant in predicting survival.